File: backend/lexer.py
import string

""" tokens string literals """
PROGRAM_LITERAL  = "program"
FUNCTION_LITERAL  = "function"
VAR_LITERAL      = "var"
IF_LITERAL       = "if"
WHILE_LITERAL    = "while"
EQUAL_LITERAL    = "equal"
NOTEQUAL_LITERAL = "notequal"
AND_LITERAL      = "and"
OR_LITERAL       = "or"
TRUE_LITERAL     = "true"
FALSE_LITERAL    = "false"
ADVANCE_LITERAL  = "advance"   # lib  
TURN_LITERAL     = "turn"      # lib  
COLOR_LITERAL    = "color"     # lib  

"""  token types  """
TT_PROGRAM    = "program"
TT_DEFINITION = "definition"
TT_KEYWORD    = "keyword"
TT_OPERATOR   = "operator"
TT_NUMBER     = "number"
TT_VAR        = "var"
TT_ID         = "id"
TT_ASSIGNMENT = "assignment"          # =  
TT_LIB        = "lib"                 # lib  
TT_BOOLEAN    = "boolean"

TT_COMMA          = "comma"          
TT_SEMICOLON      = "semicolon"       # ;  
TT_LEFT_PAREN     = "left_paren"      # (  
TT_RIGHT_PAREN    = "right_paren"     # )  
TT_LEFT_BRACKET   = "left_bracket"    # {  
TT_RIGHT_BRACKET  = "right_bracket"   # }  

LETTERS = string.ascii_letters
DIGITS = '0123456789'

class Error:
  def __init__(self, error_name, details) -> None:
    self.error_name = error_name
    self.details = details

  def __repr__(self) -> str:
    return f'Error: {self.error_name}: {self.details}'
  
  def to_string(self):
    return f'Error: {self.error_name}: {self.details}'

class Token:
  def __init__(self, type_, value=None):
    self.type = type_
    self.value = value
  
  def __repr__(self) -> str:
    if self.value:
      return f'{self.type} : {self.value}'
    return f'{self.type}'

class Lexer:
  def __init__(self, text_) -> None:
    self.pos = -1
    self.current_char = None
    self.text = ' ' + text_ + ' '
    self.advance()
  
  def advance(self):
    self.pos += 1
    self.current_char = self.text[self.pos] if self.pos < len(self.text) else None
  
  def tokenizer(self):
    tokens = []
    while self.current_char != None:
      c = self.current_char
      # stopwords
      if c in ' \n\t':
        self.advance()
      
      # keywords and identifiers
      elif c in LETTERS:
        tokens.append(self.makeLexeme())
      # numbers
      elif c in DIGITS:
        tokens.append(self.makeDigit())
      # separators
      elif c in '();{,}':
        tokens.append(self.makeSeparator())
        self.advance()
      #operators
      elif c in '+-*/':
        tokens.append(self.makeOperator())
        self.advance()
      # assignment
      elif c in '=':
        tokens.append(Token(TT_ASSIGNMENT, c))
        self.advance()
      else:
        print(Error('invalid token', c))
        self.advance()
    return tokens

  def makeLexeme(self):
    lexeme = ''
    while self.current_char in LETTERS:
      lexeme += self.current_char
      self.advance()
    if lexeme == PROGRAM_LITERAL:
      return Token(TT_PROGRAM, lexeme)

    if lexeme == WHILE_LITERAL or lexeme == IF_LITERAL:
      return Token(TT_KEYWORD, lexeme)

    if lexeme == VAR_LITERAL:
      return Token(TT_VAR, lexeme)

    if lexeme == EQUAL_LITERAL or lexeme == NOTEQUAL_LITERAL or lexeme == AND_LITERAL or lexeme == OR_LITERAL:
      return Token(TT_OPERATOR, lexeme)

    if lexeme == TRUE_LITERAL or lexeme == FALSE_LITERAL:
      return Token(TT_BOOLEAN, lexeme)

    if lexeme == ADVANCE_LITERAL or lexeme == TURN_LITERAL or lexeme == COLOR_LITERAL:
      return Token(TT_LIB, lexeme)

    return Token(TT_ID, lexeme) #identifier

  def makeDigit(self):
    num_str = ''
    while self.current_char in DIGITS:
      num_str += self.current_char
      self.advance()
    return Token(TT_NUMBER, num_str)

  def makeSeparator(self):
    c = self.current_char
    if c=='(': return Token(TT_LEFT_PAREN, c)
    if c==')': return Token(TT_RIGHT_PAREN, c)
    if c=='{': return Token(TT_LEFT_BRACKET, c)
    if c=='}': return Token(TT_RIGHT_BRACKET, c)
    if c==',': return Token(TT_COMMA, c)
    if c==';': return Token(TT_SEMICOLON, c)
    return Token("","")

  def makeOperator(self):
    return Token(TT_OPERATOR, self.current_char)



























CONSOLE_LITERAL = "console.log"

File: backend/c++/lexer.cpp
#include <bits/stdc++.h>
using namespace std;

/* 

echo "4+3 equal 7 and 4*4 notequal 16" > hola.txt & ./out < hola.txt */
/* 
python:
result = os.system('echo "4+3 equal 7 and 4*4 notequal16 16" > hola.txt & ./out < hola.txt')
 */

  /* tokens string literals */
  #define  PROGRAM_LITERAL  "program"
  #define  VAR_LITERAL      "var"
  #define  IF_LITERAL       "if"
  #define  WHILE_LITERAL    "while"
  #define  EQUAL_LITERAL    "equal"
  #define  NOTEQUAL_LITERAL "notequal"
  #define  AND_LITERAL      "and"
  #define  OR_LITERAL       "or"
  #define  TRUE_LITERAL     "true"
  #define  FALSE_LITERAL    "false"
  #define  ADVANCE_LITERAL  "advance" /* lib */
  #define  TURN_LITERAL     "turn"    /* lib */
  #define  COLOR_LITERAL    "color"   /* lib */

  /* token types  */
  #define TT_PROGRAM    "program"
  #define TT_DEFINITION "definition"
  #define TT_KEYWORD    "keyword"
  #define TT_OPERATOR   "operator"
  #define TT_NUMBER     "number"
  #define TT_VAR        "var"
  #define TT_ID         "id"
  #define TT_ASSIGNMENT "assignment" /* = */
  #define TT_LIB        "lib"
  #define TT_BOOLEAN    "boolean"

  #define TT_COMMA          "comma"         /* , */
  #define TT_SEMICOLON      "semicolon"     /* ; */
  #define TT_LEFT_PAREN     "left_paren"    /* ( */
  #define TT_RIGHT_PAREN    "right_paren"   /* ) */
  #define TT_LEFT_BRACKET   "left_bracket"  /* { */
  #define TT_RIGHT_BRACKET  "right_bracket" /* } */


struct Token
{
  string type;
  string value;
};

typedef vector<Token> Tokens;

struct Lexer 
{
  int pos;
  char current_char;
  string text;

  Lexer(string _text)
  {
    text = _text;
    text = " " + text + " ";
    pos = -1;
    advance();
  }
  Tokens tokenizer()
  {
    Tokens tokens;
    char c;
    while(pos < text.size())
    {
      c = current_char;
      if(c == '\n' || c == '\t' || c == ' ')
        advance();
      
      /* keywords and identifiers */
      else if(isalpha(c))
      {
        tokens.push_back(makeLexeme());
      }
      /* numbers */
      else if(isdigit(c))
      {
        tokens.push_back(makeDigit());
      }
      /* separators */
      else if(c=='(' || c==')' || c==';' || c==',' || c=='{' || c=='}'){
        tokens.push_back(makeSeparator());
        advance();
      }
      /* operators */
      else if(c=='+' || c=='-' || c=='*' || c=='/'){
        tokens.push_back(makeOperator());
        advance();
      }
      /* ASSIGNMENT */
      else if(c=='='){
        tokens.push_back({TT_ASSIGNMENT, string(1,c)});
        advance();
      }

      else {
        cerr << "Error: invalid token[" << c << "]" << endl;
        advance();
      }
    }

    return tokens;
  }

  Token makeLexeme()
  {
    string lexeme = "";
    while(isalpha(current_char)){
      lexeme += string(1, current_char);
      advance();
    }
    if(lexeme == PROGRAM_LITERAL)
      return {TT_PROGRAM, lexeme};
    if( lexeme == WHILE_LITERAL || lexeme == IF_LITERAL || lexeme == VAR_LITERAL)
      return {TT_KEYWORD, lexeme};
    if( lexeme == EQUAL_LITERAL || lexeme == NOTEQUAL_LITERAL || lexeme == AND_LITERAL || lexeme == OR_LITERAL)
      return {TT_OPERATOR, lexeme};
    
    if( lexeme == ADVANCE_LITERAL || lexeme == TURN_LITERAL || lexeme == COLOR_LITERAL)
      return {TT_LIB, lexeme};
    return {TT_ID, lexeme}; //identifier
  }

  Token makeDigit()
  {
    string number = "";
    while(isdigit(current_char)){
      number += string(1, current_char);
      advance();
    }
    return {TT_NUMBER, number};
  }

  Token makeSeparator()
  {
    char c = current_char;
    if(c=='(') return {TT_LEFT_PAREN, string(1, c)};
    if(c==')') return {TT_RIGHT_PAREN, string(1, c)};
    if(c=='{') return {TT_LEFT_BRACKET, string(1, c)};
    if(c=='}') return {TT_RIGHT_BRACKET, string(1, c)};
    if(c==',') return {TT_COMMA, string(1, c)};
    if(c==';') return {TT_SEMICOLON, string(1, c)};
    return {"",""};
  }

  Token makeOperator()
  {
    return {TT_OPERATOR, string(1, current_char)};
  }

  void advance()
  {
    if(pos + 1 >= text.size())
       pos = text.size();
    else {
      pos = pos + 1;
      current_char = text[pos];
    }
  }
}; //END LEXER

struct parser
{

};

/* reading file */
string readFile(string filename)
{
  ifstream in(filename, ios_base::in);
  string s = "";
  char c;
  while(!in.eof())
  {
    c = in.get();
    if(c>0)
      s += string(1, c);
  }
  return s;
}

int main(int argc, char *argv[])
{
  
  string text = readFile(argv[1]);
  Lexer l(text);
  Tokens tokens = l.tokenizer();
  cout << string(30,'.') << endl;
  for (int i = 0; i < tokens.size(); i++)
  {
    cout << tokens[i].type << ": " << tokens[i].value << endl;
  }
  

}
